File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-2.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        memo = {} # key is (index, buying or selling) -> max profit if 
        # bought curr / sold at curr
        def dfs(i, buying):
            if i >= len(prices):
                return 0
            if (i, buying) in memo:
                return memo[(i, buying)]
        
            if buying:
                buy = -prices[i] + dfs(i + 1, False)
                skip = dfs(i + 1, buying)
                res = max(buy, skip)
                memo[(i, True)] = res
                return res
            else:
                sell = prices[i] + dfs(i + 2, True)
                skip = dfs(i + 1, buying)
                res = max(sell, skip)
                memo[(i, False)] = res
                return res
        
        print(memo)
        return dfs(0, True)

File: Data Structures & Algorithms/max-water-container/submission-2.java
class Solution {
    public int maxArea(int[] heights) {
        
        int l = 0;
        int r = heights.length - 1;
        int res = 0;

        while (l < r) {
            res = Math.max(res, (r - l) * Math.min(heights[r], heights[l]));
            if (heights[r] > heights[l]) {
                l += 1;
            } else {
                r -= 1;
            }
        }

        return res;
    }
}
